### PR TITLE
[FIX] website_partner: fix website description field

### PR DESCRIPTION
**Pull request by bvr-odoo**
Steps to reproduce the bug:

- Go to the "/partners" page.
- Click on the first partner to go to their page.
- Enter "edit mode".
- Drag and drop a "Dynamic products" snippet into the right column of
the page.
- Save the page.
- Bug: the "Dynamic products" snippet is not visible on the page.

The problem is not specific to the "Dynamic products" snippet. The issue
occurs with all snippets. When the "website description" field is saved,
all HTML attributes are removed (except the "class" attribute). This
causes many bugs, which vary depending on the snippets dropped into the
field.

This commit adds the parameter "sanitize_overridable=True" to the
"website_description" field of "website_partner". It means that the
description can now be properly edited as an user with enough rights,
while keeping the same level of protection for other users editing this
field. Still buggy, below version 15.0, for a restricted editor editing
that description but should be rare.

opw-3908218

### CODE REVIEW

File: addons/website_partner/models/res_partner.py [hunk 1]
```diff
@@ -9,7 +9,7 @@ class WebsiteResPartner(models.Model):
     _name = 'res.partner'
     _inherit = ['res.partner', 'website.seo.metadata']
 
-    website_description = fields.Html('Website Partner Full Description', strip_style=True, translate=html_translate)
+    website_description = fields.Html('Website Partner Full Description', sanitize_attributes=False, translate=html_translate)
```
Comment (qsm-odoo) [inline addons/website_partner/models/res_partner.py:12]: @bvr-odoo Maybe align this with other frontend field definition: I think there is also the `sanitize_form=False` on other fields like this? And from 16.0, sanitize_overridable, to remember.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website_partner/models/res_partner.py (hunk 1 above)
Comment (bvr-odoo) [reply addons/website_partner/models/res_partner.py:12]: > sanitize_form=False

@qsm-odoo I had already taken a look at the similar "website_description" fields and it is aligned on them.

Comment (qsm-odoo) [reply addons/website_partner/models/res_partner.py:12]: I don't know which ones you looked at but most have `sanitize_form=False` like I said :wink: There are 2-3 occurrences without it and they are probably mistake as well :grimacing: But ok, let's keep it that way for now, maybe we want it that way in stable, I have to ask the security team about some things too.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@bvr-odoo In fact, this is not right: that field can be edited by the partner itself if website_slides is installed and that the partner has karma. In that case, it is just like a forum post/answer: we want full sanitization. I am not sure what solution we could have for 15.0, we can probably ignore it. From 16.0, I guess this is "fixable" by keeping the same (full) sanitization but adding the `sanitize_overridable=True` attribute. Meaning that only editor user will be able to edit that field fully and portal user will still be able to edit it, restrained by full sanitization.

Although, we should also have something like for form_sanitization: not prevent to drop anything if it won't work (just allow text). I think we already had a task for that actually? @bso-odoo maybe? You implemented that: when form sanitization is enabled, form cannot be dropped. Did not we have something to prevent dropping anything if full sanitization is enabled?

Really, it should be 2 different fields: one for the website_partner usecase and one for the website_profile usecase... probably requires deep refactoring :scream:

**Comment by qsm-odoo** [comment]
> Although, we should also have something like for form_sanitization: not prevent to drop anything if it won't work (just allow text). I think we already had a task for that actually? @bso-odoo maybe? You implemented that: when form sanitization is enabled, form cannot be dropped. Did not we have something to prevent dropping anything if full sanitization is enabled?

In fact @rdeodoo says to me we already made that part with https://github.com/odoo/odoo/pull/120588

So to confirm everything but:
- 15.0: we don't fix anything, edition of partner from an editor point of view is buggy if you start using blocks... let users upgrade to next version to handle this better.
- 16.0: we add sanitize_overridable only, meaning that the description can now be properly edited as an user with enough rights, while keeping the same level of protection for other users editing this field. Still buggy for a restricted editor editing that description but should be rare.
- 17.0: same as 16.0 but it does not appear buggy anymore, the field just cannot receive block if you are a full editor user.
- master: website_partner and website_profile should use different fields (but we can forget about that and maybe ping the related PO).

**Comment by bso-odoo** [comment]
> Although, we should also have something like for form_sanitization: not prevent to drop anything if it won't work (just allow text). I think we already had a task for that actually? @bso-odoo maybe? You implemented that: when form sanitization is enabled, form cannot be dropped. Did not we have something to prevent dropping anything if full sanitization is enabled?

Here is the PR with the initial explanation <https://github.com/odoo/odoo/pull/97398>.

Here is an attempt at re-describing it:
There are two aspects that are combined in that logic:
In the `t-snippet` definitions, blocks can be tagged as incompatible with sanitization:
See [here](https://github.com/odoo/odoo/blob/17.0/addons/web_editor/models/ir_qweb_fields.py#L85).
- `t-forbid-sanitize="true"` indicates that the block cannot be dropped in sanitized fields (e.g. embed code snippet).
- `t-forbid-sanitize="form"` indicates that the block cannot be dropped in sanitized fields, except if they explicitly do not sanitize forms.

The logic about the fields definition is [here](https://github.com/odoo/odoo/blob/17.0/addons/web_editor/models/ir_qweb_fields.py#L401).
In short:
- if the HTML field is specified with `sanitize_overridable=True` and the user has the group `base.group_sanitize_override` (by default: is a restricted editor), then none of the drop restrictions applies: the user is considered responsible/trusted enough
- if the HTML field is specified with `sanitize_attributes=True`, no `t-forbid-sanitize` drop is allowed
- if the HTML field combines `sanitize_attributes=True` but `sanitize_form=False`, drop is allowed for `t-forbid-sanitize="form"` but not for `t-forbid-sanitize="true"`

The behavior slightly evolves across versions. In 17.0, safe blocks seem to not be droppable anymore, I'll look into it.

**Comment by bso-odoo** [comment]
> I'll look into it. (task-3956565)

It was intended by task-3293600 <https://github.com/odoo/odoo/pull/120588>: `sanitize_attributes=True` makes blocks non-droppable and edition is limited to text only.

So, here is a summary for the behavior after 16.4:

- `sanitize_overridable=False, sanitize_attributes=False, sanitize_form=True` => drop safe ok, drop form blocked, drop unsafe blocked

- `sanitize_overridable=False, sanitize_attributes=False, sanitize_form=False` => drop safe ok, drop form ok, drop unsafe blocked

- `sanitize_overridable=False, sanitize_attributes=True` => no drop: text edit only

- `sanitize_overridable=True`: allow to drop anything if user is base.group_sanitize_override

File: addons/website_partner/models/res_partner.py [hunk 2]
```diff
@@ -9,7 +9,7 @@ class WebsiteResPartner(models.Model):
     _name = 'res.partner'
     _inherit = ['res.partner', 'website.seo.metadata']
 
-    website_description = fields.Html('Website Partner Full Description', strip_style=True, translate=html_translate)
+    website_description = fields.Html('Website Partner Full Description', sanitize_overridable=True, translate=html_translate)
```
Comment (qsm-odoo) [inline addons/website_partner/models/res_partner.py:12]: @bvr-odoo I don't think `strip_style` should be removed? Given my explanation https://github.com/odoo/odoo/pull/165482#issuecomment-2135528653 I still think that this is to be sanitized fully for non admin users?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

**Review by qsm-odoo — APPROVED** [review]
I push-forced to replace

> Still buggy, below version 15.0, for a restricted editor editing that description but should be rare.

with the correct explanation and more: it is still buggy for 16.0 and below (not 15.0) and ok-ish in 17.0 (nothing seems buggy there but it's not great as not properly editable).

@robodoo r+

**Comment by bvr-odoo** [comment]
@odoo/rd-security  Could we have the approval please?

**Comment by mart-e** [comment]
@qsm-odoo @bvr-odoo I am not convinced the report makes much sense but it’s safe :shrug: 
@robodoo override=ci/security
Reason: only admins/designers can bypass override
